java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/QuickOpen.java

```diff
@@ -0,0 +1,135 @@
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.CodeActionParams;
+import org.eclipse.lsp4j.MessageParams;
+import org.eclipse.lsp4j.MessageType;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.ShowDocumentParams;
+import org.netbeans.api.java.project.JavaProjectConstants;
+import org.netbeans.api.java.source.ClassIndex;
+import org.netbeans.api.java.source.ClasspathInfo;
+import org.netbeans.api.java.source.ElementHandle;
+import org.netbeans.api.java.source.ui.ElementOpen;
+import org.netbeans.api.project.Project;
+import org.netbeans.api.project.ProjectUtils;
+import org.netbeans.api.project.SourceGroup;
+import org.netbeans.modules.java.lsp.server.LspServerState;
+import org.netbeans.modules.java.lsp.server.Utils;
+import org.netbeans.modules.parsing.api.ResultIterator;
+import org.openide.filesystems.FileObject;
+import org.openide.filesystems.URLMapper;
+import org.openide.util.Lookup;
+import org.openide.util.NbBundle;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ *
+ * @author Dusan Balek
+ */
+@ServiceProvider(service = CodeActionsProvider.class)
+public class QuickOpen extends CodeActionsProvider {
+
+    public static final String QUICK_OPEN =  "java.quick.open"; // NOI18N
+    public static final String DEFAULT_PKG =  "<default package>"; // NOI18N
+    private final Gson gson = new Gson();
+
+    @Override
+    public Set<String> getCommands() {
+        return Collections.singleton(QUICK_OPEN);
+    }
+
+    @NbBundle.Messages({
+        "DN_SelectType=Select type to open",
+        "DN_NoTypeFound=No type found in openend projects"
+    })
+    @Override
+    public CompletableFuture<Object> processCommand(NbCodeLanguageClient client, String command, List<Object> arguments) {
+        LspServerState server = Lookup.getDefault().lookup(LspServerState.class);
+        if (server != null) {
+            return server.openedProjects().thenCompose(prjs -> {
+                ArrayList<QuickPickItem> items = new ArrayList<>();
+                for (Project prj : prjs) {
+                    for (SourceGroup sg : ProjectUtils.getSources(prj).getSourceGroups(JavaProjectConstants.SOURCES_TYPE_JAVA)) {
+                        FileObject root = sg.getRootFolder();
+                        for (ElementHandle<TypeElement> type : ClasspathInfo.create(root).getClassIndex().getDeclaredTypes("", ClassIndex.NameKind.PREFIX, EnumSet.of(ClassIndex.SearchScope.SOURCE))) {
+                            String qualifiedName = type.getQualifiedName();
+                            int idx = qualifiedName.lastIndexOf('.');
+                            String name = idx < 0 ? qualifiedName : qualifiedName.substring(idx + 1);
+                            String pkgName = idx < 0 ? DEFAULT_PKG : qualifiedName.substring(0, idx);
+                            items.add(new QuickPickItem(name, pkgName + " : " + root.toURI().toString(), null, false, new ElementData(type)));
```

Comment:
Wow, this will be performance - marvellous for transfer & client processing in projects like Micronaut :)) but we can only use the UI that  Microsoft has allowed for us.